sherpa/cpp_api/brpc/server.cc

```diff
@@ -0,0 +1,380 @@
+/**
+ * Copyright      2022  (authors: Pingfeng Luo)
+ *
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <chrono>  //NOLINT
+#include <condition_variable>  //NOLINT
+#include <mutex>  //NOLINT
+#include <random>
+#include <shared_mutex>
+#include <thread>  //NOLINT
+#include <vector>
+
+#include "asr.pb.h" //NOLINT
+#include "sherpa/csrc/log.h"
+#include "sherpa/csrc/online_asr.h"
+#include "sherpa/csrc/parse_options.h"
+
+#include "brpc/server.h"
+#include "brpc/stream.h"
+#include "butil/base64.h"
+#include "butil/logging.h"
+
+using namespace std::chrono_literals;  //NOLINT
+
+const int32_t MAX_ASR_RESULT = 512;
+const int32_t SAMPLE_RATE = 16000;
+const float PADDING_LEN = 0.36;
+enum StreamState : int32_t {
+  unknow_state = -1,
+  input_start,
+  endpoint_inactive,
+  endpoint_active,
+  endpoint_reset,
+  input_end,
+};
+// set endpoint_reset state in two case:
+// 1) endpoint inactive and already fetch result
+// 2) no more input
+
+// AsrDecoder
+class AsrDecoder final {
+ public:
+  AsrDecoder(const sherpa::OnlineAsrOptions opts,
+      const size_t max_batch_size = 4096,
+      const int32_t wait_ms = 10)
+    : alive_(true),
+    exit_(false),
+    decoding_(false),
+    wait_ms_(wait_ms),
+    max_batch_size_(max_batch_size),
+    online_asr_(opts) {
+      batch_size_ = std::max(1,
+          std::min(static_cast<int32_t>(max_batch_size_ >> 2), 64));
+      decoder_ = std::thread(&AsrDecoder::batch_decode, this);
+      decoder_.detach();
+      srand(42);
+      warmup_decode();
+    }
+
+  ~AsrDecoder() {
+    while (!exit_) {
```

Comment:
Instead of using polling, I suggest that we can use
`std::future<void>` and `std::promise<void>` here.
You can set the promise in the `decoder_` thread.

----

Alternatively, we don't call `decoder_.detach()` in the constructor but use `decoder_.join()` here in the destructor.